player_darwin.go

```diff
@@ -0,0 +1,97 @@
+package oto
+
+// #cgo CFLAGS: -I/System/Library/Frameworks
+// #cgo LDFLAGS: -framework CoreAudio -framework AudioToolbox -framework Foundation
+// #include "player_darwin.h"
+import "C"
+
+// global buffer since it must be accessed from the go_input_callback
+var Buf []byte
+
+type player struct {
+	audioPlayer C.AudioPlayer
+	playing     bool
+}
+
+func (p *player) Write(b []byte) (n int, err error) {
+	if !p.playing {
+		p.playing = true
+		p.StartPlayback()
+	}
+	Buf = append(Buf, b...)
+	return len(b), nil
+}
+
+func (p *player) Close() error {
+	C.ClosePlayer(&p.audioPlayer)
+	return nil
+}
+
+func NewDarwinPlayer() (*player, error) {
+	return newPlayer(0, 0, 0, 0)
+}
+
+func newPlayer(sampleRate, channelNum, bytesPerSample, bufferSizeInBytes int) (*player, error) {
+	p := &player{}
+
+	//bufferSize := bufferSizeInBytes / (channelNum * bytesPerSample)
+	//buf = make([]byte, 0, bufferSize)
+	p.audioPlayer = C.NewAudioPlayer(C.Float64(sampleRate), C.UInt32(channelNum), C.UInt32(bytesPerSample*8))
+	return p, nil
+}
+
+func (p *player) StartPlayback() {
+	C.StartPlayback(&p.audioPlayer)
+}
+
+func (p *player) stopPlayback() {
+	p.playing = false
+	C.StopPlayback(&p.audioPlayer)
+}
+
+// go_input_callback is responsible for placing information onto the underlying C circular buffer to be consumed by
+// the go_output_callback, which will play audio to the underlying device
+
+//export go_input_callback
+func go_input_callback(player *C.AudioPlayer,
+	flags *C.AudioUnitRenderActionFlags,
+	audioTimeStamp *C.AudioTimeStamp,
+	inBusNumber C.UInt32,
+	inNumFrames C.UInt32,
+	ioData *C.AudioBufferList) C.OSStatus {
+
+	//j := player.startingFrameCount
+	//cycleLength := float32(44100.0 / 440.0)
```

Comment:
Yeah, like before, this is in no way final and I will be removing things that are commented out.